### PR TITLE
Use request-scoped MCP clients for Responses API tools

### PR DESCRIPTION
**Pull request by zhaowenzi**
## Problem
The PR identifies a security vulnerability: the connection pool caches by URL (server_key), but auth tokens are NOT part of that key.

User A: URL=X, token=A → creates connection, cached as "X"
User B: URL=X, token=B → gets cached connection "X" with token A's auth\!

## PR Solution

To address the security issue where the connection pool is cached only by URL (`server_key`) and does not include authentication context, we considered the following two approaches:

---

### Option 1: Per-request MCP Tool Context

**Approach**  
Create an isolated MCP tool context per request. Each request builds and uses its own connection context, ensuring that authentication state is never shared across users.

**Pros**
- Easy to implement  
- Strong isolation at the request level  
- Completely avoids cross-user credential leakage  

**Cons**
- Introduces additional overhead (connection setup / teardown per request)  
- Loses most connection pooling benefits  

---

### Option 2: Auth-aware Connection Caching

**Approach**  
Extend the connection pool cache key to include authentication information, so that connections are reused only when both the URL and auth context match.

**Pros**
- Allows connection reuse for the same user with the same authentication  
- Reduces connection creation overhead compared to per-request isolation  

**Cons**
- In enterprise MCP servers, authentication is almost always required  
- With the upcoming support for custom headers, using only URL + auth token as the cache key is insufficient — headers must also be included, since not all MCP servers follow the standard authentication mechanism, and many rely on custom headers (e.g. x-api-key) for authentication.
- Once headers are added to the cache key, the key space becomes highly fragmented  
- As a result, the practical benefit of connection pooling becomes very limited, while complexity increases significantly  

---

### Conclusion

Given the security guarantees required and the limited real-world benefits of connection reuse when authentication and headers are fully accounted for, **Option 1 (Per-request MCP tool context)** is the preferred solution.  

It provides clear security boundaries with minimal implementation risk, at the cost of some additional per-request overhead, which is considered acceptable for correctness and safety.

## Notes
- gRPC paths remain unchanged, will add this in a separate PR
- OpenAI Responses does not currently support static MCP tools, so RequestMcpContext intentionally ignores static servers/tools for now; we can revisit if/when builtin tools need it

### CODE REVIEW

**Comment by zhaowenzi** [comment]
Hi @slin1237, as we discussed earlier, this PR builds a request-scoped MCP context. Please let me know if the implementation is the right direction. The gRPC path is not covered yet; I can open a follow‑up PR for that later.

**Comment by slin1237** [comment]
## PR Review: Request-scoped MCP clients

### Problem Identified 

The PR correctly identifies a **security vulnerability**: the connection pool caches by URL (`server_key`), but auth tokens are NOT part of that key.

```
User A: URL=X, token=A → creates connection, cached as "X"
User B: URL=X, token=B → gets cached connection "X" with token A's auth\!
```

### PR Solution Assessment

The `RequestMcpContext` approach is **correct** and fixes the security issue:
- Fresh connections per request via `connect_request_client`
- Private `ToolInventory` per request
- Proper cleanup in `Drop`

---

### Architectural Concerns

**1. Logic leakage to router layer**

`build_request_mcp_context` in `mcp_utils.rs:72-167` contains MCP domain logic:
- Server config creation
- Transport type detection (`/sse` heuristic)
- Connection management
- Inventory loading

This should live in `McpManager`, not the router layer.

**2. Code duplication**

Nearly identical logic exists in both:
- `ensure_request_mcp_client` (old, lines 84-163)
- `build_request_mcp_context` (new, lines 72-167)

**3. No caching benefit**

Every request creates new connections even if the same user with the same auth makes multiple requests (slight performance cost).

---

### Suggested Better Architecture

**Option 1: Move context creation into McpManager (Recommended)**

```rust
// src/mcp/manager.rs
impl McpManager {
    /// Create a request-scoped MCP context from response tools.
    /// 
    /// This method:
    /// 1. Parses MCP tools from the request
    /// 2. Creates fresh connections (not pooled) to preserve per-request auth
    /// 3. Returns a self-contained context with its own inventory
    pub async fn create_request_context(
        &self,
        tools: Option<&[ResponseTool]>,
    ) -> Option<Arc<RequestMcpContext>> {
        let tools = tools?;
        
        let mut clients = HashMap::new();
        let inventory = Arc::new(ToolInventory::new());
        let mut server_keys = Vec::new();
        let mut has_mcp_tools = false;

        for tool in tools {
            if \!matches\!(tool.r#type, ResponseToolType::Mcp) || tool.server_url.is_none() {
                continue;
            }
            has_mcp_tools = true;
            
            let server_config = self.parse_tool_server_config(tool)?;
            let server_key = Self::server_key(&server_config);
            
            match self.connect_request_client(&server_config).await {
                Ok(client) => {
                    Self::load_server_inventory(&inventory, &server_key, &client).await;
                    clients.insert(server_key.clone(), client);
                    if \!server_keys.contains(&server_key) {
                        server_keys.push(server_key);
                    }
                }
                Err(err) => {
                    warn\!("Failed to connect MCP server {}: {}", server_key, err);
                }
            }
        }

        if \!has_mcp_tools || clients.is_empty() {
            return None;
        }

        Some(Arc::new(RequestMcpContext::new(inventory, clients, server_keys)))
    }
    
    /// Parse ResponseTool into McpServerConfig
    fn parse_tool_server_config(&self, tool: &ResponseTool) -> Option<McpServerConfig> {
        let server_url = tool.server_url.as_ref()?.trim().to_string();
        
        if \!server_url.starts_with("http://") && \!server_url.starts_with("https://") {
            warn\!("Ignoring MCP server_url with unsupported scheme: {}", server_url);
            return None;
        }

        let name = tool.server_label.clone().unwrap_or_else(|| "request-mcp".to_string());
        let token = tool.authorization.clone();

        let transport = if server_url.contains("/sse") {
            McpTransport::Sse { url: server_url, token }
        } else {
            McpTransport::Streamable { url: server_url, token }
        };

        Some(McpServerConfig {
            name,
            transport,
            proxy: None,
            required: false,
        })
    }
}
```

**Router becomes simple:**

```rust
// src/routers/openai/responses/non_streaming.rs
let active_mcp = mcp_manager
    .create_request_context(original_body.tools.as_deref())
    .await;
```

**Option 2: Auth-aware caching (future optimization)**

If you want connection reuse for the same user with same auth:

```rust
impl McpManager {
    /// Generate cache key including auth hash for per-user caching
    fn server_key_with_auth(config: &McpServerConfig) -> String {
        let base_key = Self::server_key(config);
        let auth_hash = match &config.transport {
            McpTransport::Sse { token, .. } | McpTransport::Streamable { token, .. } => {
                token.as_ref()
                    .map(|t| format\!("{:x}", md5::compute(t.as_bytes())))
                    .unwrap_or_default()
            }
            _ => String::new(),
        };
        if auth_hash.is_empty() {
            base_key
        } else {
            format\!("{}:{}", base_key, auth_hash)
        }
    }
}
```

This allows caching while maintaining auth isolation: `https://mcp.example.com:abc123` vs `https://mcp.example.com:def456`.

---

### Summary

| Aspect | Current PR | Suggested |
|--------|-----------|-----------|
| Security fix | ✅ Correct | ✅ Same |
| Encapsulation | ⚠️ Logic in router | ✅ Logic in McpManager |
| Code duplication | ⚠️ Two similar functions | ✅ Single source of truth |
| Caching | ❌ No reuse | Optional: auth-aware caching |
| Testability | ⚠️ Router-coupled | ✅ McpManager-isolated |

The PR is **mergeable as-is** for the security fix, but I'd recommend refactoring the context creation logic into `McpManager` to properly encapsulate MCP concerns and avoid router-layer leakage.

File: src/routers/mcp_utils.rs [hunk 1]
```diff
@@ -70,6 +70,98 @@ pub fn extract_server_label(tools: Option<&[ResponseTool]>, default_label: &str)
...
+    if clients.is_empty() {
+        warn!("All MCP request tools failed to connect");
+        return None;
+    }
+
+    Some(Arc::new(RequestMcpContext::new(
+        inventory,
+        clients,
+        server_keys,
+    )))
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply src/routers/mcp_utils.rs]: Done

File: src/mcp/manager.rs [hunk 2]
```diff
@@ -778,6 +788,104 @@ impl McpManager {
...
+        let mut tools = Vec::new();
+
+        for (_tool_name, server_key, tool_info) in self.inventory.list_tools() {
+            if server_keys.contains(&server_key) {
+                tools.push(tool_info);
+            }
+        }
+
+        tools
+    }
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply src/mcp/manager.rs:924]: Done

File: src/mcp/manager.rs [hunk 3]
```diff
@@ -778,6 +788,104 @@ impl McpManager {
...
+                            if let Err(err) = client.cancel().await {
+                                warn!("Error closing request client: {}", err);
+                            }
+                        });
+                    }
+                    Err(_) => {
+                        warn!("Request MCP client still has active references on drop");
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply src/mcp/manager.rs]: Done

File: src/routers/mcp_utils.rs [hunk 4]
```diff
@@ -70,6 +70,98 @@ pub fn extract_server_label(tools: Option<&[ResponseTool]>, default_label: &str)
...
+                            server_keys.push(server_key);
+                        }
+                    }
+                    Err(err) => {
+                        warn!(
+                            "Failed to connect MCP request server {}: {}",
+                            server_key, err
+                        );
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply src/routers/mcp_utils.rs]: Done

File: src/mcp/manager.rs [hunk 5]
```diff
@@ -778,6 +788,104 @@ impl McpManager {
...
+    pub fn list_tools_for_servers(&self, server_keys: &[String]) -> Vec<Tool> {
+        let mut tools = Vec::new();
+
+        for (_tool_name, server_key, tool_info) in self.inventory.list_tools() {
+            if server_keys.contains(&server_key) {
+                tools.push(tool_info);
+            }
+        }
+
+        tools
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply src/mcp/manager.rs]: Done

File: src/mcp/manager.rs [hunk 6]
```diff
@@ -778,6 +788,104 @@ impl McpManager {
...
+impl Drop for RequestMcpContext {
+    /// Best-effort cleanup for request-scoped clients.
+    fn drop(&mut self) {
+        if self.clients.is_empty() {
+            return;
+        }
+
+        if let Ok(handle) = tokio::runtime::Handle::try_current() {
+            let clients: Vec<Arc<McpClient>> = self.clients.values().cloned().collect();
+            for client in clients {
```
[the remark this answers is not in the document]
Comment (zhaowenzi) [reply src/mcp/manager.rs]: Done

**Comment by zhaowenzi** [comment]
Thanks! @slin1237 

- moved request context creation into McpManager::create_request_context
- added a private parse_tool_server_config helper for URL/transport parsing
- removed build_request_mcp_context from router utils
- OpenAI Responses now only calls mcp_manager.create_request_context(...)
- ensure_request_mcp_client is still kept for the gRPC path (as noted in the PR description); I’ll align gRPC with the new request-scoped flow in a follow-up PR.

**Comment by slin1237** [comment]
## Detailed Rust Code Review

### 1. Inline paths should be imports

The full paths are repeated multiple times:
```rust
crate::protocols::responses::ResponseTool
crate::protocols::responses::ResponseToolType::Mcp
```

Add to imports at top of file:
```rust
use crate::protocols::responses::{ResponseTool, ResponseToolType};
```

---

### 2. Fragile SSE detection heuristic

```rust
let transport = if server_url.contains("/sse") {  // ⚠️ matches "/message/sse-test"
```

`contains("/sse")` could false-positive on URLs like `https://example.com/message/sse-handler`. Consider:
```rust
let transport = if server_url.ends_with("/sse") || server_url.contains("/sse?") {
```

Or document why the current heuristic is acceptable.

---

### 3. `call_tool` early return pattern

The current pattern has extra nesting:
```rust
if let Some((server_key, tool_info)) = self.inventory.get_tool(tool_name) {
    // ... lots of code ...
    return client.call_tool(request).await...;
}
Err(McpError::ToolNotFound(tool_name.to_string()))
```

More idiomatic:
```rust
let (server_key, tool_info) = self.inventory
    .get_tool(tool_name)
    .ok_or_else(|| McpError::ToolNotFound(tool_name.to_string()))?;

let tool_schema = Some(serde_json::Value::Object((*tool_info.input_schema).clone()));
// ... rest without extra nesting ...
```

---

### 4. `server_keys` field duplicates `clients.keys()`

```rust
pub struct RequestMcpContext {
    inventory: Arc<ToolInventory>,
    clients: HashMap<String, Arc<McpClient>>,
    server_keys: Vec<String>,  // always == clients.keys()
}
```

Options:
1. Remove field, derive on demand (slight perf cost)
2. If insertion order matters, document this or use `IndexMap`
3. Keep as-is if the clone cost at creation is acceptable

---

### 5. Silent failure in `Drop` when no runtime

```rust
if let Ok(handle) = tokio::runtime::Handle::try_current() {
    // cleanup
}
// else: silently leaks connections
```

Consider logging the failure case:
```rust
} else {
    warn!("No tokio runtime available for MCP client cleanup");
}
```

---

### 6. `has_mcp_tools` naming

```rust
let Some(server_config) = self.parse_tool_server_config(tool) else {
    continue;
};
has_mcp_tools = true;
```

`parse_tool_server_config` returns `None` for non-MCP tools AND for MCP tools with invalid URLs. So `has_mcp_tools` really means "found at least one valid MCP tool config". Minor naming nitpick.

---

### 7. No unit tests for new code

Consider adding tests for:
- `parse_tool_server_config` edge cases (invalid URLs, missing fields)
- `create_request_context` partial failure scenarios
- `list_tools_for_servers` filtering logic

---

### Summary

| Issue | Severity | 
|-------|----------|
| Inline paths | Low (style) |
| SSE heuristic | Medium (potential false positive) |
| Early return pattern | Low (style) |
| Redundant `server_keys` | Low (data duplication) |
| Silent Drop failure | Low (debuggability) |
| Naming | Low (clarity) |
| No unit tests | Medium (maintainability) |

Overall the code is solid. The SSE heuristic is the main item worth a second look.

**Comment by zhaowenzi** [comment]
> Detailed Rust Code Review

Done

**Comment by slin1237** [comment]
please also change the PR description for what the problems are and the solution
So the history can be trackers
